clients/src/main/java/org/apache/kafka/common/log/remote/metadata/storage/RocksDBMetadataStore.java

```diff
@@ -0,0 +1,171 @@
+import java.nio.file.Files;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+
+import static org.apache.kafka.common.config.ConfigDef.Importance.MEDIUM;
+import static org.apache.kafka.common.config.ConfigDef.Range.atLeast;
+import static org.apache.kafka.common.config.ConfigDef.Type.INT;
+
+public class RocksDBMetadataStore implements MetadataStore {
+    private static String DB_DIR = "remote-segment-metadata";
+
+    public static class Config extends AbstractConfig {
+        public static final String REMOTE_METADATA_ROCKSDB_CACHE_SIZE_MB_PROP = "remote.log.metadata.rocksdb.cache.size.mb";
+        public static final String REMOTE_METADATA_ROCKSDB_CACHE_SIZE_MB_DOC = "Remote segment metadata store RocksDB in-memory cache size in MB";
+        private static final ConfigDef CONFIG_DEF;
+
+        static {
+            CONFIG_DEF = new ConfigDef().define(REMOTE_METADATA_ROCKSDB_CACHE_SIZE_MB_PROP, INT, 50, atLeast(1), MEDIUM, REMOTE_METADATA_ROCKSDB_CACHE_SIZE_MB_DOC);
+        }
+
+        public Config(Map<?, ?> originals) {
+            super(CONFIG_DEF, originals);
+        }
+    }
+
+    public static void loadLibrary() {
+        RocksDB.loadLibrary();
+    }
+
+    RocksDB db;
+    Cache cache;
+    RLSMSerDe.RLSMSerializer serializer = new RLSMSerDe.RLSMSerializer();
+    RLSMSerDe.RLSMDeserializer deserializer = new RLSMSerDe.RLSMDeserializer();
+
+    RocksDBMetadataStore(String logDir, Map<String, ?> configs) {
+        Config conf = new Config(configs);
+
+        cache = new LRUCache(conf.getInt(Config.REMOTE_METADATA_ROCKSDB_CACHE_SIZE_MB_PROP));
+        final Options options = new Options();
+        options.setCreateIfMissing(true);
+        options.setTableFormatConfig(new BlockBasedTableConfig().setBlockCache(cache));
+        File dir = new File(logDir + File.separator + DB_DIR);
+        try {
+            Files.createDirectories(dir.getAbsoluteFile().toPath());
+            db = RocksDB.open(options, dir.getAbsolutePath());
+        } catch(IOException | RocksDBException ex) {
+            throw new KafkaException("Failed to initialize RocksDB: " + ex.getMessage(), ex);
+        }
+    }
+
+    @Override
+    public void flush() throws IOException {
+        FlushOptions options = new FlushOptions();
+        options.setWaitForFlush(true);
+        try {
+            db.flush(options);
+        } catch(RocksDBException ex) {
+            throw new KafkaException("Failed to persist remote segment metadata to local storage: " + ex.getMessage(), ex);
+        }
+    }
+
+    @Override
+    public void load() throws IOException {
+    }
+
+    @Override
+    public void update(TopicPartition tp, RemoteLogSegmentMetadata metadata) {
+        try {
+            WriteBatch batch = new WriteBatch();
+            final NavigableMap<Long, RemoteLogSegmentId> segmentIds = getSegmentIds(tp);
```

Comment:
Deserializing and building a map for every get or update may be costly. Did you think about searching based on prefix based key instead of realizing all the entries and build a map?

Reply:
We also need to check whether this model causes high contention as we have a single DB instance at brokers for all the topic partition updates it receives from rlmm topic.

Reply:
For each topic-partition, the segment id map / list is accessed in 2 cases
1. Metadata update, when a new segment is shipped to remote storage or a segment is deleted.
2. When a consumer tries to consume a remote segment of topic-partition

Depends on the configuration, 1 happens every several minutes to a couple of hours. 2 only happens when the consumers is consuming very old data. Either case does not happen very frequently. So the performance shouldn't be a problem.

As long as the data is cached in memory, serializing / deserializing should be pretty fast (comparing with serializing / deserializing the corresponding Kafka request / response messages, and network latency)

We can also split the remote-segment-list into smaller pieces. So that the append only happens on the last piece, and delete only happens on the first piece. But, when we looking up an offset, we still need an index to search for an offset, and can find out the "next" offset of a given offset. This means we will need an index for the offsets. If the index is also stored in RocksDB, the data structure would be something like a B+tree.

Reply:
ser/des may not be an issue but creating transient data like map instances etc may cause pressure on minor GC. 

RLMM on any broker can be subscribed to a large no remote log segment metadata partitions in a cluster and it may process those events. This should be tested out. We can discuss/explore different approaches once we run perf test on this standalone store with large ingress of messages onto the remote log segment metadata topic.